### PR TITLE
Update documentation in README

### PR DESCRIPTION
**Pull request by mpvosseller**
- Updates to latest syntax / api
- Updates to prefer enums that start with lowercase letters
- Added documentation on using the new `authorizationHeader` property

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
 ```swift
 public struct TokenConfiguration: Configuration {
     public let accessToken: String?
-    public let apiEndpoint = "https://my.webservice.example/api/2.0/"
+    public let apiEndpoint = "https://my.webservice.example/api/2.0/"
```
Comment (pietbrauer) [inline README.md]: Is the whitespace needed here?

File: README.md [hunk 2]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
...
 We will need something that connects the router and the configuration to make provide a convenient interface. The common way of doing this is to use a `struct` or a `class` that does it for you.
 
 ```swift
+struct User : Codable {
+}
+
 struct MyWebservice {
     var configuration: Configuration
-
+
```
Comment (pietbrauer) [inline README.md]: Please remove the whitespace

File: README.md [hunk 3]
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
 
 ```swift
 enum MyRouter: Router {
-    case GetMyself(Configuration)
-
+    case getMyself(Configuration)
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 4]
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
...
+    case getMyself(Configuration)
+    
     var configuration: Configuration {
         switch self {
-        case .GetMyself(let config): return config
+        case .getMyself(let config): return config
         }
     }
-
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 5]
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
...
+    
     var method: HTTPMethod {
         switch self {
-        case .GetMyself:
+        case .getMyself:
             return .GET
         }
     }
-
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 6]
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
...
     var encoding: HTTPEncoding {
         switch self {
-        case .GetMyself:
-            return .URL
+        case .getMyself:
+            return .url
         }
     }
-
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 7]
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
...
     var path: String {
         switch self {
-        case .GetMyself:
+        case .getMyself:
             return "myself"
         }
     }
-
-    var params: [String: AnyObject] {
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 8]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
...
+}
+```
+
+In the above `Configuration` the `accessToken` will be passed as a URL parameter named `access_token` with each request. Alternatively you can have the `accessToken` passed in an HTTP Authorization header by setting the `authorizationHeader` property to the desired token type. As an example the following `Configuration` passes it as a Bearer token.
 
+```swift
+public struct TokenConfiguration: Configuration {
+    public let accessToken: String?
+    public let apiEndpoint = "https://my.webservice.example/api/2.0/"
+    public let authorizationHeader: String? = "Bearer"
```
Comment (pietbrauer) [inline README.md:97]: We can omit the `: String?` here.

File: README.md [hunk 9]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
...
+    public let errorDomain: String = "com.my.customErrorDomain"
```
Comment (pietbrauer) [inline README.md]: We can omit the `: String` here.

File: README.md [hunk 10]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
...
+        
     public init(_ accessToken: String? = nil) {
         self.accessToken = accessToken
     }
 }
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 11]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
...
 ```
 
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

File: README.md [hunk 12]
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
...
-
+    
     init(configuration: Configuration) {
         self.configuration = configuration
     }
-
-    func getMyself(session: RequestKitURLSession = NSURLSession.sharedSession(), completion: (response: Response<[String: AnyObject]>) -> Void) -> RequestKitURLSession {
-        let router = MyRouter.GetMyself(configuration)
-        router.loadJSON(session, expectedResultType: [String: AnyObject].self) { json, error in
+
```
Comment (pietbrauer) [inline README.md]: Please remove the extra whitespace.

**Review by pietbrauer — CHANGES_REQUESTED** [review]
Nice work! I prefer no trailing whitespace but else this looks very very good.

File: README.md (hunk 8 above)
Comment (mpvosseller) [reply README.md:97]: Frustratingly it appears that we can't omit the type here. I'm not sure if this is the expected Swift behavior or if I'm doing something dumb or if this is a Swift bug? I put together some code you can throw into a playground to demo the issue:
https://gist.github.com/mpvosseller/02c6bbf853442c5de3c2fc9adc139e00

**Comment by mpvosseller** [comment]
OK I fixed the whitespace, removed a few unneeded type declarations, and also fixed a minor typo. 

The last remaining issue is that (shockingly) it does appear that an implementation of the `Configuration` protocol needs to explicitly declare `authorizationHeader` as an optional String in its implementation. 

So like this:
```
public let authorizationHeader: String? = "Bearer"
```
or like this
```
var authorizationHeader : String? {
    return "BEARER"
}
```

Unfortunately failing to do so results in the default implementation getting called which returns `nil`.

It is pretty annoying and I wonder if it is a Swift bug. Here is some code to demo the issue:
https://gist.github.com/mpvosseller/02c6bbf853442c5de3c2fc9adc139e00

If we are weary of doing this, the best alternative I can think of is to make this a non-optional String (with some default value like `BEARER`) that can be overridden and then add a new protocol method which is what we use to decide whether or not to use it like `shouldUseAuthorizationHeader`

WDYT?

**Comment by mpvosseller** [comment]
After simplifying the code in this gist a bit I suspect this is probably the expected behavior from Swift.
https://gist.github.com/mpvosseller/02c6bbf853442c5de3c2fc9adc139e00

**Comment by mpvosseller** [comment]
I had a conversation with Swift engineers on Twitter about this type issue. It appears to be a known issue which the may add compiler warnings for. I also filed a radar for the issue.
https://twitter.com/mpv/status/968938220225474560

The options I see are:

**Option 1** - Leave it as is and RequestKit users must explicitly declare their `authorizationHeader` implementation as `String?` and if they fail to do so the compiler (for now) won't complain and RequestKit just won't use the header. In this case we should update the docs to explain this.

**Option 2** - Make `authorizationHeader` a non optional `String` that defaults to `Bearer` and add a new protocol property like `shouldUseAuthorizationHeader` that returns a `Bool` and returns `false` by default.

WDYT? Any other options?

**Comment by pietbrauer** [comment]
I would opt for Option 1

**Review by pietbrauer — APPROVED** [review]

**Pull request merged**